### PR TITLE
Emoji Count command(ec)

### LINKED ISSUE #454: `.emojicount` command for displaying the number of emojis categorically
## Description

An `.emojicount` command can be used to display the categorical emoji count in the server through an embed.

## Reasoning

Due to recent changes as to how emojis in the server in the server are categorized with prefixes, implementing such a command would be fairly trivial.

## Proposed Implementation

An `.emojicount` command can be invoked with zero or more categories. For every category requested, each line in the embed would consist of the following format:
```python
[randomly_chosen_emoji_in_category] [name_of_the_category]: [number_of_emojis_in_category]
```
The resulting embed could be prettified as well, and I'm looking for any suggestions.

## Would you like to implement this yourself?

- [x] I'd like to implement this feature myself
- [x] Anyone can implement this feature

### PR DESCRIPTION
**Pull request by Anubhav1603**
"Closes #454 "

command take 1 category name or many or none and it gives embed with emoji count and random emoji, if you give valid category and invalid category together bot will replace the invalid category and shows the output but if u give only invalid category bot will show the error 

command invoke with 1 category
category
command invoke with more than 1 category
category1
error embed
error

- [x] Join the [**Python Discord Community**](https://discord.gg/python)?
- [x] If dependencies have been added or updated, run `pipenv lock`?
- [x] **Lint your code** (`pipenv run lint`)?
- [x] Set the PR to **allow edits from contributors**?

### CODE REVIEW

**Comment by Akarys42** [comment]
Hey @Anubhav1603, thanks for the PR! You seem to have a failed linting, please make sure to run `pipenv run precommit` on your local copy. 
```
bot/exts/evergreen/emoji_count.py:42:32: ANN001 Missing type annotation for function argument 'ctx'
bot/exts/evergreen/emoji_count.py:55:26: ANN001 Missing type annotation for function argument 'ctx'
bot/exts/evergreen/emoji_count.py:65:24: ANN001 Missing type annotation for function argument 'ctx'
```

File: bot/exts/evergreen/emoji_count.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+
+class EmojiCount(commands.Cog):
+    """Command that give random emoji based on category."""
+
+    def __init__(self, bot: commands.Bot):
+        self.bot = bot
+
+    def embed_builder(self, emoji: dict) -> discord.Embed:
+        """Genrates embed with emoji name and count."""
```
Comment (Akarys42) [inline bot/exts/evergreen/emoji_count.py:21]:
```suggestion
        """Generates an embed with the emoji names and count."""
```

File: bot/exts/evergreen/emoji_count.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        if len(emoji) == 1:
+            for key, value in emoji.items():
+                print(key)
+                embed.description = f"There are **{len(value)}** emojis in the **{key}** category"
+                embed.set_thumbnail(url=random.choice(value).url)
+        else:
+            msg = ''
+            for key, value in emoji.items():
+                emoji_choice = random.choice(value)
+                error_msg = f'There are **{len(value)}** emojis in the **{key}** category\n'
```
Comment (Akarys42) [inline bot/exts/evergreen/emoji_count.py:35]: Why is this variable called `error_msg`? It doesn’t sound like an error message to me.

File: bot/exts/evergreen/emoji_count.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+                msg += f'<:{emoji_choice.name}:{emoji_choice.id}> {error_msg}'
+            embed.description = msg
+        log.trace('Emoji embed sent')
```
Comment (Akarys42) [inline bot/exts/evergreen/emoji_count.py:38]: It doesn’t get send here, something like `'Emoji embed built'` would make more sense, or move it where it is actually sent.

File: bot/exts/evergreen/emoji_count.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        return embed
+
+    @staticmethod
+    def generate_invalid_embed(ctx: commands.Context) -> discord.Embed:
+        """Genrates error embed."""
+        embed = discord.Embed()
+        embed.color = Colours.soft_red
+        embed.title = "Invalid Input"
```
Comment (Akarys42) [inline bot/exts/evergreen/emoji_count.py:46]: We usually use `bot.constants.ERROR_REPLIES` for that.

File: bot/exts/evergreen/emoji_count.py [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        emoji_dict = {}
+        for emoji in ctx.guild.emojis:
+            emoji_dict.update({emoji.name.split("_")[0]: []})
+        error_comp = ', '.join(key for key in emoji_dict.keys())
+        embed.description = f"These are the valid categories\n```{error_comp}```"
+        log.trace("Error embed sent")
```
Comment (Akarys42) [inline bot/exts/evergreen/emoji_count.py:52]: Again, the embed isn’t send here.

File: bot/exts/evergreen/emoji_count.py [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+        log.trace("Emoji dict sent")
+        return emojis
+
+    @commands.command(name="ec")
+    async def ec(self, ctx: commands.Context, *, emoji: str = None) -> Optional[str]:
+        """Returns embed with emoji category and info given by user."""
+        emoji_dict = {}
+        for a in ctx.guild.emojis:
+            if emoji is None:
+                log.trace("Emoji Category doesn't provided by the user")
```
Comment (Akarys42) [inline bot/exts/evergreen/emoji_count.py:70]:
```suggestion
                log.trace("Emoji Category not provided by the user")
```

File: bot/exts/evergreen/emoji_count.py [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+    @staticmethod
+    def generate_invalid_embed(ctx: commands.Context) -> discord.Embed:
+        """Genrates error embed."""
+        embed = discord.Embed()
+        embed.color = Colours.soft_red
+        embed.title = "Invalid Input"
+        emoji_dict = {}
+        for emoji in ctx.guild.emojis:
+            emoji_dict.update({emoji.name.split("_")[0]: []})
```
Comment (Akarys42) [inline bot/exts/evergreen/emoji_count.py:49]: This is a lot more clearer, please replace all the similar calls by this. 
```suggestion
            emoji_dict[emoji.name.split("_")[0]] = []
```

File: bot/exts/evergreen/emoji_count.py [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        error_comp = ', '.join(key for key in emoji_dict.keys())
+        embed.description = f"These are the valid categories\n```{error_comp}```"
+        log.trace("Error embed sent")
+        return embed
+
+    def emoji_list(self, ctx: commands.Context, emojis: dict) -> Dict:
```
Comment (Akarys42) [inline bot/exts/evergreen/emoji_count.py:55]: This whole function seems pretty inefficient, it looks like you can replace it by something like that (note that I replaced `emojis` to be a list of the categories you want) : 
```py
def emoji_list(ctx, categories):
    out = {category: [] for category in categories}

    for emoji in ctx.guild.emojis:
        category = emoji.name.split('_')[0]
        if category in out:
            out[category].append(emoji)
    return out
```

File: bot/exts/evergreen/emoji_count.py [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+        """Genrates dictionary of emojis given by the user."""
+        for emoji in ctx.guild.emojis:
+            for key, value in emojis.items():
+                if emoji.name.split("_")[0] == key:
+                    value.append(emoji)
+        log.trace("Emoji dict sent")
```
Comment (Akarys42) [inline bot/exts/evergreen/emoji_count.py:61]: Again, the dictionary is being built here, not sent.

File: bot/exts/evergreen/emoji_count.py [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+        return emojis
+
+    @commands.command(name="ec")
```
Comment (Akarys42) [inline bot/exts/evergreen/emoji_count.py:64]: The fully command name should be `emoji_count` with `ec` being an alias.

**Review by Akarys42 — CHANGES_REQUESTED** [review]
Your PR has some pretty good points, but they are some little things to change.

File: bot/exts/evergreen/emoji_count.py (hunk 2 above)
Comment (Anubhav1603) [reply bot/exts/evergreen/emoji_count.py:35]: oopsy

File: bot/exts/evergreen/emoji_count.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+
+    def __init__(self, bot: commands.Bot):
+        self.bot = bot
+
+    def embed_builder(self, emoji: dict) -> discord.Embed:
+        """Generates an embed with the emoji names and count."""
+        embed = discord.Embed()
+        embed.color = Colours.orange
+        embed.title = "Emoji Count"
+        embed.timestamp = datetime.datetime.utcnow()
```
Comment (ks129) [inline bot/exts/evergreen/emoji_count.py:25]:
```suggestion
        embed = discord.Embed(
            color=Colours.orange,
            title="Emoji Count",
            timestamp=datetime.datetime.utcnow()
        )
```

File: bot/exts/evergreen/emoji_count.py [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+                embed.description = f"There are **{len(value)}** emojis in the **{key}** category"
+                embed.set_thumbnail(url=random.choice(value).url)
+        else:
+            msg = ''
+            for key, value in emoji.items():
+                emoji_choice = random.choice(value)
+                emoji_info = f'There are **{len(value)}** emojis in the **{key}** category\n'
+                msg += f'<:{emoji_choice.name}:{emoji_choice.id}> {emoji_info}'
+            embed.description = msg
+        log.trace('Emoji embed built')
```
Comment (ks129) [inline bot/exts/evergreen/emoji_count.py:37]: This is pretty pointless logging

File: bot/exts/evergreen/emoji_count.py [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+        """Genrates error embed."""
+        embed = discord.Embed()
+        embed.color = Colours.soft_red
+        embed.title = random.choice(ERROR_REPLIES)
+        emoji_dict = {}
+        for emoji in ctx.guild.emojis:
+            emoji_dict[emoji.name.split("_")[0]] = []
+        error_comp = ', '.join(key for key in emoji_dict.keys())
+        embed.description = f"These are the valid categories\n```{error_comp}```"
+        log.trace("Error embed built")
```
Comment (ks129) [inline bot/exts/evergreen/emoji_count.py:51]: This is too pointless logging

File: bot/exts/evergreen/emoji_count.py [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+            embed.description = msg
+        log.trace('Emoji embed built')
+        return embed
+
+    @staticmethod
+    def generate_invalid_embed(ctx: commands.Context) -> discord.Embed:
+        """Genrates error embed."""
+        embed = discord.Embed()
+        embed.color = Colours.soft_red
+        embed.title = random.choice(ERROR_REPLIES)
```
Comment (ks129) [inline bot/exts/evergreen/emoji_count.py:45]:
```suggestion
        embed = discord.Embed(
            color=Colours.soft_red,
            title=random.choice(ERROR_REPLIES)
        )
```

File: bot/exts/evergreen/emoji_count.py [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+        emoji_dict = {}
+        for emoji in ctx.guild.emojis:
+            emoji_dict[emoji.name.split("_")[0]] = []
```
Comment (ks129) [inline bot/exts/evergreen/emoji_count.py:48]: Add empty newline after this line.

File: bot/exts/evergreen/emoji_count.py [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+
+    def emoji_list(self, ctx: commands.Context, categories: dict) -> Dict:
+        """Generates an embed with the emoji names and count."""
+        out = {category: [] for category in categories}
+
+        for emoji in ctx.guild.emojis:
+            category = emoji.name.split('_')[0]
+            if category in out:
+                out[category].append(emoji)
+        log.trace("Emoji dict built")
```
Comment (ks129) [inline bot/exts/evergreen/emoji_count.py:62]: This is too pointless logging

File: bot/exts/evergreen/emoji_count.py [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+            if emoji is None:
+                log.trace("Emoji Category not provided by the user")
+                emoji_dict.update({a.name.split("_")[0]: []})
+            elif a.name.split("_")[0] in emoji:
+                log.trace("Emoji Category provided by the user")
+                emoji_dict.update({a.name.split("_")[0]: []})
+        emoji_dict = self.emoji_list(ctx, emoji_dict)
+        if len(emoji_dict) == 0:
+            embed = self.generate_invalid_embed(ctx)
+            log.trace("Error embed received")
```
Comment (ks129) [inline bot/exts/evergreen/emoji_count.py:79]: This is too pointless logging

File: bot/exts/evergreen/emoji_count.py [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+        else:
+            embed = self.embed_builder(emoji_dict)
+            log.trace("Emoji embed received")
```
Comment (ks129) [inline bot/exts/evergreen/emoji_count.py:82]: This is also pointless logging

File: bot/exts/evergreen/emoji_count.py [hunk 19]
```diff
@@ -0,0 +1,89 @@
...
+        await ctx.send(embed=embed)
+        log.trace("Embed sent")
```
Comment (ks129) [inline bot/exts/evergreen/emoji_count.py:84]: This is too pointless logging.

**Review by ks129 — CHANGES_REQUESTED** [review]
Logging should show something useful. Embed sent isn't good logging. Also could you make code to smaller blocks?

File: bot/exts/evergreen/emoji_count.py (hunk 14 above)
Comment (Anubhav1603) [reply bot/exts/evergreen/emoji_count.py:45]: is it make much difference?

File: bot/exts/evergreen/emoji_count.py (hunk 19 above)
Comment (Anubhav1603) [reply bot/exts/evergreen/emoji_count.py:84]: thats y i am using trace for logging

File: bot/exts/evergreen/emoji_count.py (hunk 14 above)
Comment (ks129) [reply bot/exts/evergreen/emoji_count.py:45]: This is current style. Other cogs use same style for embed building.

File: bot/exts/evergreen/emoji_count.py (hunk 19 above)
Comment (ks129) [reply bot/exts/evergreen/emoji_count.py:84]: Even when trace, for developers this add a lot of log messages to console, what maybe makes like error searching harder.

Comment (Anubhav1603) [reply bot/exts/evergreen/emoji_count.py:84]: ok

**Review by Akarys42 — APPROVED** [review]
It would be nice if you could add some empty lines in the code to make it a bit less dense, but overhaul this PR is on point 👌

**Review by ks129 — APPROVED** [review]
Looks good

**Pull request merged**